### ISSUE TITLE
VS Code OpenShift Toolkit extension does not detect connection on the cluster it is deployed on

### ISSUE THREAD

**Issue by apupier** [eclipse/che#23183]
### Describe the bug

VS Code OpenShift Toolkit extension does not detect connection on the cluster it is deployed on even if VS Code Kubernetes is able to detect it

### Che version

7.92@latest

### Steps to reproduce

1. open workspace with VS Code OpenShift tolkit extension, for instance https://github.com/apupier/reproducer-che-vscode-openshift
2. Trust authors
3. Wait for the install of the extensions
4. Open the Kuebrnetes perspective and notice the connection available to the local OpenShift cluster:
![image](https://github.com/user-attachments/assets/5ffd0cea-442c-41d6-89bb-5441d71afa99)
5. Open the OpenShift perspective and notice that no connection is available:
![image](https://github.com/user-attachments/assets/bf87f15e-dc9f-460a-95d4-a1d60de849f1)

### Expected behavior

to have connection visible in VS Code OpenShift perspective

### Runtime

OpenShift

### Installation method

other (please specify in additional context)

### Environment

Dev Sandbox (workspaces.openshift.com)

**Comment by ibuziuk** [comment]
@adietish could you please take a look ?

**Comment by adietish** [comment]
@vrubezhny is looking at it in https://github.com/redhat-developer/vscode-openshift-tools/issues/4535.

**Comment by vrubezhny** [comment]
@apupier Could you tell, please, if this issue is a regression? 
I mean, did it work as expected until `vscode-openshift-tools` v.1.14.0 was released or did it never detect the connection in the cluster it runs on?

**Comment by apupier** [comment]
I'm not sure.
I remember it was connecting automatically several months (years?) ago but given that it was long time ago I'm not sure if I was using VS code Kubernetes only or if I was using VS Code OpenShift Toolkit.

**Comment by vrubezhny** [comment]
The screenshots you posted show different extensions - the first one (logged in) is from Kubernates extension, while the second one (not logged in) is from Openshift Tools extension. I haven't seen how Openshift Tools extension was behaving when running on the cluster (whether it was logged in the same way as K8s extension or not), but if K8s extension was configuring somehow the "current" cluster we're running on into the "current-context" property in Kube config then Openshift extension could pick that new context after it was created.
Starting from v.1.14.0 Openshift extension doesn't depend on/required K8s extension to be installed - so on the first run (using your reproducer example) I see   "...configuration cannot be loaded..." error instead of being logged in:

file:///home/jeremy/Pictures/Screenshots/Screenshot%20from%202024-11-21%2023-10-44.png

**Comment by apupier** [comment]
based on this command in the terminal:
```
reproducer-che-vscode-openshift (master) $ ldd --version
ldd (GNU libc) 2.28
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```
it seems that the base image is containg a too old version glibc compared to what the oc and odo binaries are supporting

**Comment by vrubezhny** [comment]
@apupier Do we have any idea on how we can change/update the base image so it has the required version of GLIBC?

`oc`/`odo` are the binaries we embed and use as is and we need these binaries to be fully functional. And this problem could be the one of the reason we cannot login to the cluster we're running on (actually to any cluster).

**Comment by vrubezhny** [comment]
Yes, I've checked - `alizer`, `oc`, `odo` CLI binaries we use require GLIBC 2.32+/2.34+ while `func` and `helm` CLI binaries work don't (or work well with the provided lib version) - the binaries of the latest versions are bundled together with the OSTools extension, so the extension fails to work as expected.
Kubernetes extension looks like is using the `oc` CLI binary found in user's PATH (which is compatible)  - so it works well (at least it shows what cluster it is logged in to).

**Comment by vrubezhny** [comment]
Depends on the https://github.com/eclipse-che/che/issues/23034 to be solved.

**Comment by vrubezhny** [comment]
@apupier  Do you have `~/.kube/config` existing?  Could you check this, please?

**Comment by apupier** [comment]
> @apupier Do you have `~/.kube/config` existing? Could you check this, please?

It seems I have no ~/.kube/config available. The VS Code Kubernetes extension is not finding the connection neither this time

**Comment by vrubezhny** [comment]
OK, then it looks like https://github.com/redhat-developer/vscode-openshift-tools/issues/3872 is still reproducible. 

PS: VS Openshift Toolkit doesn't depend on VS Kubernetes extension, so you don't need to add the last one to your `extension.json`. But you may still need to have it to check how do the extensions work in regard of logging in and/or other functions.

**Comment by vrubezhny** [comment]
But, all I can do here is to create a `~/.kube/config` with a minimalistic contents (or report an error if it cannot be created), so the extension can be successfully activated.
See: https://github.com/redhat-developer/vscode-openshift-tools/pull/4676

However I wouldn't (even if I can) define any `clusters`/`users`/`contexts` properties (especially pointing to a cluster and user currently logged in - this preferably is to be done during the workspace configuration and before the extension starts.

**Comment by vrubezhny** [comment]
For the UDI 9 image, it looks like the system environment variables aren't set properly... At least `$HOME`:

![image](https://github.com/user-attachments/assets/2ce09a98-23e2-45f8-b816-79ab0461afa3)

**Comment by vrubezhny** [comment]
The VSCode Openshift Toolkit activation doesn't fail anymore when using the v.1.18.0...

... but still, when trying to open CHE workspace, I have `$HOME = /`  and no Kube config created, so Openshift Toolkit can't show any information on the cluster it's connected to nor watch for the Kube config changes. The same goes to the Kubernetes extension as well.

**Comment by vrubezhny** [comment]
@ibuziuk Problem is that starting the CHE workspace using the @apupier's example (a bit modified) I don't get `$HOME` variable set properly (it suddenly resolves to `/`)  and I don't have any `/home/user/.kube` folder created:

![Image](https://github.com/user-attachments/assets/5ccb41b6-d88b-41b2-8b89-c6a89b859e34)

**Comment by ibuziuk** [comment]
@vrubezhny well, in your sample the image does not have oc available e.g. https://github.com/che-incubator/quarkus-api-example/pull/54 seems to work fine

@apupier could you please review and let us know if the issue can be resolved?

**Comment by vrubezhny** [comment]
@adietish @ibuziuk @apupier Thanks! 

VSCode Openshift Tools works as expected after I modified @apupier's reproducer example with the the image and some other container properties from [quarkus-api-example](https://github.com/che-incubator/quarkus-api-example/pull/54) suggested by @ibuziuk.  

![Image](https://github.com/user-attachments/assets/278f9fa4-9d02-4c4e-b5d2-7b2bdbfca0a1)

The resulting reproducer example is used to start the ^^^ workspace:  https://github.com/vrubezhny/reproducer-che-vscode-openshift/blob/master/readme.md 
 
So, I believe, this issue can be closed as resolved now.

**Comment by apupier** [comment]
If I understand well, the image provided by default on the Red Hat developer sandbox is not configured to have the connection automatically created?

When using  https://github.com/apupier/reproducer-che-vscode-openshift , which is using the default devfile, the connection is not available.
When using https://github.com/vrubezhny/reproducer-che-vscode-openshift/blob/master/readme.md , which is using a custom devfile, the connection is available.

What is the reason to prevent it working out of the box?

**Comment by ibuziuk** [comment]
@vrubezhny indeed, it does not seem to work with http://workspaces.openshift.com/#https://github.com/apupier/reproducer-che-vscode-openshift 

<img width="1601" alt="Image" src="https://github.com/user-attachments/assets/ebf81c07-9afe-4e88-bebb-ca3014bb929b" />

@adietish there is still a problem on the extension end I believe

**Comment by adietish** [comment]
@ibuziuk looks like it 😢 . Will point @vrubezhny to it when he gets back next week.

**Comment by vrubezhny** [comment]
@adietish @ibuziuk IMHO, That's not a RHEL9-based image Workspace. At least it still has GLIBC v.2.28...

![Image](https://github.com/user-attachments/assets/ea285a25-66b3-406e-a822-d6b371c8f540)

... while the `OpenShift Client` binary requires it to be at least of version 2.32:

![Image](https://github.com/user-attachments/assets/89e0f206-4bc5-47c5-9c80-7d10a9841509)

**Comment by vrubezhny** [comment]
@apupier @adietish @ibuziuk  When `quay.io/devfile/universal-developer-image:ubi9-latest` image is used - everything works well ([see the test example here](https://github.com/vrubezhny/reproducer-che-vscode-openshift/blob/master/readme.md))

![Image](https://github.com/user-attachments/assets/105ddc49-55f6-4c89-9077-cbd69bb344ae)

**Comment by vrubezhny** [comment]
> If I understand well, the image provided by default on the Red Hat developer sandbox is not configured to have the connection automatically created?

That's not only about the connection. `vscode-openshift-tools` needs GLIBC >= v.2.32 for `Openshift CLI (oc)` to make it possible to be executed. It won't work with UBI8 (RHEL8-based) images
So, I expect the default devfile to be changed to use an UBI9 image

**Comment by apupier** [comment]
> So, I expect the default devfile to be changed to use an UBI9 image

based on https://github.com/redhat-developer/devspaces-images/commit/1da57d328b07662e19ddc50168c5a0d55830fdd6 i would expect to have UBI9 already. Wondering why the UBI 9 is not already the default one

**Comment by ibuziuk** [comment]
UBI 9 is going to be the default as part of the https://github.com/eclipse-che/che/issues/23321 cc @martinszuc
@vrubezhny however, I do not quite grok why it does not work with UBI 8 - https://github.com/eclipse-che/che/issues/23183#issuecomment-2755153665

On the screenshot, you can see that HOME, kubeconfig, etc, are all set. Why the extension can not process it correctly?

**Comment by apupier** [comment]
> [@vrubezhny](https://github.com/vrubezhny) however, I do not quite grok why it does not work with UBI 8 - [#23183 (comment)](https://github.com/eclipse-che/che/issues/23183#issuecomment-2755153665)
> 
> On the screenshot, you can see that HOME, kubeconfig, etc, are all set. Why the extension can not process it correctly?

it is because the oc binary embedded requires a newer version of glibc than what is available in RHEL 8

**Comment by vrubezhny** [comment]
> UBI 9 is going to be the default as part of the [#23321](https://github.com/eclipse-che/che/issues/23321) cc [@martinszuc](https://github.com/martinszuc)
> [@vrubezhny](https://github.com/vrubezhny) however, I do not quite grok why it does not work with UBI 8 - [#23183 (comment)](https://github.com/eclipse-che/che/issues/23183#issuecomment-2755153665)

Yea, as @apupier answered: `oc` binary requires GLIBC to be of version >= 2.32, while UBI8-image has it only of version 2.28

**Comment by vrubezhny** [comment]
Depends on https://github.com/eclipse-che/che/issues/23321

**Comment by ibuziuk** [comment]
Closing, should be fixed on developer sandbox
